service/configinit.go

```diff
@@ -0,0 +1,546 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package service // import "go.opentelemetry.io/collector/service"
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/AlecAivazis/survey/v2"
+	"github.com/AlecAivazis/survey/v2/terminal"
+	"github.com/spf13/cobra"
+	"go.uber.org/zap"
+	"gopkg.in/yaml.v3"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/config"
+	"go.opentelemetry.io/collector/confmap"
+	"go.opentelemetry.io/collector/confmap/provider/yamlprovider"
+	"go.opentelemetry.io/collector/service/internal/components"
+)
+
+const (
+	sampleConfigHeader = `Opentelemetry Collector sample configuration. Double check the pipeline configuration
+by ensuring the desired components belong to the respective pipeline type(ex metrics, logs, traces).`
+)
+
+type configCommand struct {
+	exiter    func(args ...error) error
+	stdio     terminal.Stdio
+	available *componentsAvailable
+}
+
+type componentsAvailable struct {
+	exporters  []string
+	processors []string
+	receivers  []string
+	extensions []string
+	pipelines  []string
+}
+
+type optionsSelected struct {
+	Exporters  []string
+	Processors []string
+	Receivers  []string
+	Extensions []string
+	Pipelines  []string
+
+	ConfigFileName string
+	DryRun         bool
+}
+
+var defaultConfigFileName = "otelcol.config.yaml"
+
+var cfgCmd *configCommand
+var log *zap.Logger
+var opts = new(optionsSelected)
+
+func init() {
```

Comment:
I don't see a strong reason to use an init function here. Perhaps name it differently and invoke it where the logger needs to be set?